lib/ElementHandle.js

```diff
@@ -137,24 +151,64 @@ class ElementHandle extends JSHandle {
     return {x, y, width, height};
   }
 
+  /**
+   * @return {!Promise<?{x: number, y: number, width: number, height: number}>}
+   */
+  async _assertBoundingBox() {
+    const boundingBox = await this.boundingBox();
+    if (!boundingBox)
+      throw new Error('Node is not visible');
+    return boundingBox;
+  }
+
   /**
    *
    * @param {!Object=} options
    * @returns {!Promise<Object>}
    */
   async screenshot(options = {}) {
-    await this._scrollIntoViewIfNeeded();
+    let needsViewportReset = false;
+
+    let boundingBox = await this._assertBoundingBox();
+
+    const viewport = Object.assign({}, this._page.viewport());
+
+    const newViewport = {
+      width: Math.max(viewport.width, Math.ceil(boundingBox.width)),
+      height: Math.max(viewport.height, Math.ceil(boundingBox.height)),
+    };
+
+    if (newViewport.width !== viewport.width || newViewport.height !== viewport.height) {
+      await this._page._emulationManager.emulateViewport(this._client, newViewport);
+
+      needsViewportReset = true;
+    }
+
+    await this.executionContext().evaluate(function(el) {
+      const box = el.getBoundingClientRect();
+
+      window.scrollTo(
+          box.left + window.scrollX,
+          box.top + window.scrollY
+      );
+    }, this);
+
+    boundingBox = await this._assertBoundingBox();
```

Comment:
reseting the bounding box and the above scrolling code can go inside the same `if` where we set the new viewport. That way we don't do send extra messages into chrome for elements that fit on screen.

Reply:
I tried it out, but the scroll logic often needs to occur even if we don't resize the viewport.

With the currently pushed code:

![screenshot-element-off-screen](https://user-images.githubusercontent.com/936069/35245597-15d9bece-ff81-11e7-8a52-b463bef62469.png)

With scroll conditional on resizing viewport (where no viewport resize occurs)

![screenshot-element-off-screen](https://user-images.githubusercontent.com/936069/35245466-c19669fc-ff80-11e7-8c43-08adbb13524d.png)

In this example, the viewport is 500x500. The top of the element we're screenshotting is at y: 400. Without the scroll, it's only screenshotting at max the first 500x500 of the page.


Full test case for completeness
```js
    it('should capture full element when off screen', async({page, server}) => {
      // compare with .to-screenshot size
      await page.setViewport({width: 500, height: 500});

      await page.setContent(`
        something above
        <style>div.spacer {
          border: 2px solid red;
          background: red;
          height: 400px;
          width: 50px;
        }
        div.to-screenshot {
          border: 2px solid blue;
          background: rgba(0, 0, 0, 0.5);
          width: 400px;
          height: 200.5px;
          margin-left: 50px;
          transform: scaleY(1.2);
        }
        ::-webkit-scrollbar {
          display: none;
        }
        </style>
        <div class="spacer"></div>
        <div class="to-screenshot"></div>
        <div class="spacer"></div>
      `);

      await page.evaluate(function() {
        window.scrollTo(0, 0);
      });

      const elementHandle = await page.$('div.to-screenshot');
      const screenshot = await elementHandle.screenshot();
      expect(screenshot).toBeGolden('screenshot-element-off-screen.png');
    });
```

Reply:
That makes sense. I didn't realize you weren't scrolling at the top. Trading an extra boundingClientRect for a scroll is a good trade. Leave things how they are.